bob-access/src/credentials.rs

```diff
@@ -59,64 +74,201 @@ impl CredentialsKind {
+            ResolveState::InProgress => false,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DeclaredCredentials {
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+    kind: CredentialsKind,
+}
+
+impl DeclaredCredentials {
+    pub fn internode_builder(node_name: impl Into<String>) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::InterNode(node_name.into()))
+    }
+
+    pub fn token_builder(token: impl Into<String>) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Token(token.into()))
+    }
+
+    pub fn userpass_builder(username: impl Into<String>, password: impl Into<String>,) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        })
+    }
+
+    pub fn ip(&self) -> &Vec<SocketAddr> {
+        &self.address
+    }
+
+    pub fn replace_addresses(&mut self, addresses: Vec<SocketAddr>) {
+        self.address = addresses;
+    }
+
+    pub fn hostname(&self) -> &Option<String> {
+        &self.hostname
+    }
+
+    pub fn validate_internode(&self) -> bool {
+        (!self.address.is_empty() || self.hostname.is_some()) &&
+        self.kind.is_internode()
+    }
+
+    pub fn kind(&self) -> &CredentialsKind {
+        &self.kind
+    }
+}
+
+#[derive(Debug)]
+pub struct DeclaredCredentialsBuilder {
+    kind: CredentialsKind,
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+}
+
+impl DeclaredCredentialsBuilder {
+    fn new(kind: CredentialsKind) -> Self {
+        Self {
+            kind,
+            address: Vec::new(),
+            hostname: None,
+        }
+    }
+
+    pub fn with_address(mut self, address: SocketAddr) -> Self {
+        self.address = vec![address];
+        self
+    }
+
+    pub fn with_hostname(mut self, hostname: String) -> Self {
+        self.hostname = Some(hostname);
+        self
+    }
+
+    pub fn build(self) -> DeclaredCredentials {
+        DeclaredCredentials {
             address: self.address,
-            kind: self.kind.clone(),
+            kind: self.kind,
+            hostname: self.hostname,
+        }
+    }
+}
+
+
+#[derive(Debug)]
+pub struct DCredentialsResolveGuard {
+    credentials: DeclaredCredentials,
+    resolve_state: ResolveState,
+    resolve_threshold: Instant,
+    resolve_protection_period_ms: u64,
+}
+
+impl DCredentialsResolveGuard {
+    pub fn new(credentials: DeclaredCredentials, resolve_protection_period_ms: u64) -> Self {
+        Self {
+            credentials,
+            resolve_state: ResolveState::Resolved(0),
+            resolve_threshold: Instant::now() + Duration::from_millis(resolve_protection_period_ms),
+            resolve_protection_period_ms,
         }
     }
+
+    pub fn update_resolve_state(&mut self, authenticated: bool) -> bool {
+        if self.credentials.hostname().is_none() {
```

Comment:
I suggest to declare a function in `DeclaredCredentials` that returns true if the address resolved from hostname (`fn is_address_resolved_from_hostname(..) -> bool`)